src/mame/machine/k573mcr.cpp

```diff
@@ -1,24 +1,485 @@
+  |        8.25 MHz                  CN67 |
+  |                       EP4M16          |
+  |             DRAM4M                    |
+  |                                       |
+  |                               --------|
+  | USB-A                    CN64 |
+  |                               |
+  |          ADM485JR             |
+  | USB-B                    CN63 |
+  |-------------------------------|
 
-/*
-  GE885-PWB(A)A ( contains Toshiba tmpr3904af, ram, rom, tranceiver and glue ).
+Notes:
+	DIP8       - 8-position DIP switch
+	CN61       - BS8PSHF1AA 8 pin connector, connects to memory card harness
+	CN62       - BS8PSHF1AA 8 pin connector
+	CN63       - 6P-SHVQ labeled "0", GE885-JB security dongle is connected here
+	CN64       - 6P-SHVQ labeled "1"
+	CN65       - B4PS-VH, 4 pin power connector
+	CN67       - BS15PSHF1AA, 15-pin connector, unpopulated
+	USB-A      - USB-A connector
+	USB-B      - USB-B connector, connects to USB on System 573 motherboard
+	ADM485JR   - Analog Devices ADM485 low power EIA RS-485 transceiver
+	TMPR3904AF - Toshiba TMPR3904AF RISC Microprocessor
+	XCS05/10   - XILINX XCS10XL VQ100AKP9909 A2026631A
+	DRAM4M     - Silicon Magic 66 MHz C9742 SM81C256K16CJ-35, 256K x 16 EDO DRAM
+	EP4M16     - ROM labeled "855-A01"
 */
 
+#include "emu.h"
+#include "k573mcr.h"
+
 k573mcr_device::k573mcr_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
-	device_t(mconfig, KONAMI_573_MEMORY_CARD_READER, tag, owner, clock)
+	jvs_device(mconfig, KONAMI_573_MEMORY_CARD_READER, tag, owner, clock),
+	m_controllers(*this, "controllers"),
+	m_meta(*this, "META")
 {
+	m_ram = std::make_unique<uint8_t[]>(RAM_SIZE);
 }
 
 void k573mcr_device::device_start()
 {
+	jvs_device::device_start();
+
+	save_pointer(NAME(m_ram), RAM_SIZE);
+	save_item(NAME(m_is_memcard_initialized));
+	save_item(NAME(m_status));
+	save_item(NAME(m_psx_clock));
+	save_item(NAME(m_psx_rx_data));
+	save_item(NAME(m_psx_rx_bit));
+}
+
+void k573mcr_device::device_reset()
+{
+	jvs_device::device_reset();
+
+	std::fill_n(m_ram.get(), RAM_SIZE, 0);
+	m_is_memcard_initialized = false;
+	m_status = 0;
+	m_psx_clock = false;
+	m_psx_rx_data = 0;
+	m_psx_rx_bit = 0;
+}
+
+void k573mcr_device::device_add_mconfig(machine_config &config)
+{
+	// The actual controllers are only used in Guitar Freaks and it was
+	// meant to be used with the home version PS1 guitar controller.
+	// The PS1 guitar controller isn't emulated in MAME and it doesn't
+	// make much sense to have it enabled by default. You can still select
+	// a controller through the slots menu.
+	// The memory card ports are still usable even without a controller
+	// enabled which is the main reason for using the PSX controller ports.
+	PSXCONTROLLERPORTS(config, m_controllers, 0);
+	m_controllers->rxd().set(FUNC(k573mcr_device::write_rxd));
+	PSX_CONTROLLER_PORT(config, "port1", psx_controllers, nullptr);
+	PSX_CONTROLLER_PORT(config, "port2", psx_controllers, nullptr);
```

Comment:
This allows the user to select the connected devices.  Since this is a sealed-in-a-cabinet system, would it be better to have fixed peripherals set in machine configuration?  You can add `.set_fixed(true)` to each of them to not expose them as user-selectable options.

Reply:
I think technically allowing the user to select the connected devices is more accurate. Whether or not the game will read it is another story, you can connect any memory card and controller device to the machines. The same machine plate with the memory card and controller ports are used on all supported machines.

Picture:
![image](https://user-images.githubusercontent.com/63495610/104326264-e3944880-552c-11eb-9164-b7f911dc5357.png)

Reply:
Oh, I didn’t realise the controller ports were actually on the front panel like that, I thought it was more like the Neo Geo MVS variants with console-style controller ports inside the machine.  My bad.